File: projects/06-object-store/bench/haystack_small/main.rs
//! FileCas vs Haystack — small-object packing microbench.
//!
//! In-process: temp data dir + [`Store`] write/read only (no HTTP). Measures
//! commit + open latency/throughput and on-disk file count for thousands of
//! tiny unique blobs. See `bench/haystack_small/README.md`.
//!
//! ```text
//! cargo run --release -p object-store --features bench-tools --bin haystack_small
//! ```

use std::path::{Path, PathBuf};
use std::time::{Duration, Instant};

use chrono::Utc;
use object_store::object::Digest;
use object_store::store::{BlobLayoutKind, Store};
use serde::Serialize;
use tempfile::TempDir;
use tokio::io::AsyncReadExt;

struct Config {
    count: u32,
    size: u64,
    warmup: u32,
    drop_caches: bool,
}

impl Config {
    fn from_env() -> Self {
        Self {
            count: env_u32("COUNT", 10_000),
            size: parse_size_token(&std::env::var("SIZE").unwrap_or_else(|_| "4K".into()))
                .unwrap_or_else(|| {
                    eprintln!("bad SIZE — use e.g. 256, 1K, 4K, 16K");
                    std::process::exit(2);
                }),
            warmup: env_u32("WARMUP", 100),
            drop_caches: std::env::var("DROP_CACHES").ok().as_deref() == Some("1"),
        }
    }
}

fn env_u32(name: &str, default: u32) -> u32 {
    std::env::var(name)
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or(default)
}

fn parse_size_token(tok: &str) -> Option<u64> {
    let tok = tok.trim().to_uppercase();
    let (num, mult) = if let Some(n) = tok.strip_suffix('K') {
        (n, 1024u64)
    } else if let Some(n) = tok.strip_suffix('M') {
        (n, 1024 * 1024)
    } else if let Some(n) = tok.strip_suffix('G') {
        (n, 1024 * 1024 * 1024)
    } else {
        (tok.as_str(), 1)
    };
    num.parse::<u64>().ok().map(|n| n * mult)
}

/// Unique payload so CAS dedup cannot short-circuit FileCas.
fn unique_payload(size: usize, index: u32) -> Vec<u8> {
    let mut out = vec![0u8; size];
    let mut state = 0xDEAD_BEEF_u64 ^ (index as u64) ^ ((size as u64) << 17);
    for b in &mut out {
        state = state.wrapping_mul(6364136223846793005).wrapping_add(1);
        *b = (state >> 33) as u8;
    }
    // Guarantee uniqueness even if LCG collided somehow: stamp the index.
    let stamp = index.to_le_bytes();
    let n = stamp.len().min(out.len());
    out[..n].copy_from_slice(&stamp[..n]);
    out
}

fn shuffle_indices(n: usize, seed: u64) -> Vec<usize> {
    let mut idx: Vec<usize> = (0..n).collect();
    let mut state = seed;
    for i in (1..n).rev() {
        state = state.wrapping_mul(6364136223846793005).wrapping_add(1);
        let j = (state as usize) % (i + 1);
        idx.swap(i, j);
    }
    idx
}

fn percentile_ms(sorted: &[Duration], p: f64) -> f64 {
    if sorted.is_empty() {
        return 0.0;
    }
    let idx = ((p * (sorted.len() as f64 - 1.0)).round() as usize).min(sorted.len() - 1);
    sorted[idx].as_secs_f64() * 1000.0
}

fn maybe_drop_caches(enabled: bool) {
    if !enabled {
        return;
    }
    match std::process::Command::new("sudo")
        .args(["-n", "sh", "-c", "echo 3 > /proc/sys/vm/drop_caches"])
        .status()
    {
        Ok(s) if s.success() => eprintln!("  dropped page caches"),
        Ok(s) => eprintln!("  drop_caches exited {s} — continuing with warm cache"),
        Err(e) => eprintln!("  drop_caches failed ({e}) — continuing with warm cache"),
    }
}

fn human_bytes(n: u64) -> String {
    const K: f64 = 1024.0;
    let n = n as f64;
    if n < K {
        format!("{n:.0} B")
    } else if n < K * K {
        format!("{:.1} KiB", n / K)
    } else if n < K * K * K {
        format!("{:.2} MiB", n / (K * K))
    } else {
        format!("{:.2} GiB", n / (K * K * K))
    }
}

#[derive(Debug, Serialize)]
struct Footprint {
    objects_files: u64,
    volume_files: u64,
    objects_bytes: u64,
    volume_bytes: u64,
    index_bytes: u64,
}

fn count_files_and_bytes(dir: &Path) -> (u64, u64) {
    let mut files = 0u64;
    let mut bytes = 0u64;
    let Ok(entries) = std::fs::read_dir(dir) else {
        return (0, 0);
    };
    let mut stack: Vec<PathBuf> = entries.filter_map(|e| e.ok().map(|e| e.path())).collect();
    while let Some(path) = stack.pop() {
        let Ok(meta) = path.metadata() else {
            continue;
        };
        if meta.is_dir() {
            if let Ok(rd) = std::fs::read_dir(&path) {
                for e in rd.flatten() {
                    stack.push(e.path());
                }
            }
        } else if meta.is_file() {
            files += 1;
            bytes += meta.len();
        }
    }
    (files, bytes)
}

fn measure_footprint(root: &Path) -> Footprint {
    let objects = root.join("objects");
    let volumes = root.join("volumes");
    let (objects_files, objects_bytes) = count_files_and_bytes(&objects);
    let mut volume_files = 0u64;
    let mut volume_bytes = 0u64;
    let mut index_bytes = 0u64;
    if let Ok(rd) = std::fs::read_dir(&volumes) {
        for e in rd.flatten() {
            let path = e.path();
            let Ok(meta) = path.metadata() else {
                continue;
            };
            if !meta.is_file() {
                continue;
            }
            let name = path.file_name().and_then(|n| n.to_str()).unwrap_or("");
            if name.ends_with(".dat") {
                volume_files += 1;
                volume_bytes += meta.len();
            } else if name == "needles.json" {
                index_bytes = meta.len();
            }
        }
    }
    Footprint {
        objects_files,
        volume_files,
        objects_bytes,
        volume_bytes,
        index_bytes,
    }
}

#[derive(Serialize)]
struct PhaseStats {
    samples: u32,
    wall_secs: f64,
    ops_per_sec: f64,
    latency_p50_ms: f64,
    latency_p99_ms: f64,
}

#[derive(Serialize)]
struct LayoutReport {
    layout: String,
    write: PhaseStats,
    read: PhaseStats,
    footprint: Footprint,
}

#[derive(Serialize)]
struct Report {
    created_at: String,
    count: u32,
    size_bytes: u64,
    warmup: u32,
    drop_caches: bool,
    layouts: Vec<LayoutReport>,
}

fn summarize_phase(samples: &[Duration], wall: Duration) -> PhaseStats {
    let mut sorted = samples.to_vec();
    sorted.sort();
    let n = samples.len() as f64;
    let wall_secs = wall.as_secs_f64();
    PhaseStats {
        samples: samples.len() as u32,
        wall_secs,
        ops_per_sec: if wall_secs > 0.0 { n / wall_secs } else { 0.0 },
        latency_p50_ms: percentile_ms(&sorted, 0.50),
        latency_p99_ms: percentile_ms(&sorted, 0.99),
    }
}

async fn run_layout(kind: BlobLayoutKind, cfg: &Config) -> LayoutReport {
    let dir = TempDir::new().expect("temp data dir");
    let store = Store::open_with_layout(dir.path(), kind).expect("open store");
    let size = cfg.size as usize;

    eprintln!(
        "→ layout={}  count={}  size={}",
        kind.as_str(),
        cfg.count,
        human_bytes(cfg.size)
    );

    // Warmup commits (not timed).
    for i in 0..cfg.warmup {
        let payload = unique_payload(size, u32::MAX - i);
        store.commit_bytes(&payload).await.expect("warmup commit");
    }

    maybe_drop_caches(cfg.drop_caches);

    let mut digests: Vec<Digest> = Vec::with_capacity(cfg.count as usize);
    let mut write_samples = Vec::with_capacity(cfg.count as usize);
    let write_wall = Instant::now();
    for i in 0..cfg.count {
        let payload = unique_payload(size, i);
        let t0 = Instant::now();
        let digest = store.commit_bytes(&payload).await.expect("commit");
        write_samples.push(t0.elapsed());
        digests.push(digest);
    }
    let write_wall = write_wall.elapsed();
    let write = summarize_phase(&write_samples, write_wall);
    eprintln!(
        "  write: {:.1} ops/s  p50={:.3}ms  p99={:.3}ms  wall={:.2}s",
        write.ops_per_sec, write.latency_p50_ms, write.latency_p99_ms, write.wall_secs
    );

    let footprint = measure_footprint(dir.path());
    eprintln!(
        "  footprint: objects_files={}  volume_files={}  objects={}  volumes={}  needles.json={}",
        footprint.objects_files,
        footprint.volume_files,
        human_bytes(footprint.objects_bytes),
        human_bytes(footprint.volume_bytes),
        human_bytes(footprint.index_bytes),
    );

    maybe_drop_caches(cfg.drop_caches);

    let order = shuffle_indices(digests.len(), 0xC0FFEE ^ cfg.count as u64);
    let mut read_samples = Vec::with_capacity(digests.len());
    let mut buf = Vec::with_capacity(size);
    let read_wall = Instant::now();
    for &i in &order {
        let digest = &digests[i];
        let t0 = Instant::now();
        let mut reader = store.open_blob(digest).await.expect("open");
        buf.clear();
        reader.read_to_end(&mut buf).await.expect("read");
        read_samples.push(t0.elapsed());
        assert_eq!(
            buf.len(),
            size,
            "byte length for digest {}",
            digest.as_str()
        );
    }
    let read_wall = read_wall.elapsed();
    let read = summarize_phase(&read_samples, read_wall);
    eprintln!(
        "  read:  {:.1} ops/s  p50={:.3}ms  p99={:.3}ms  wall={:.2}s",
        read.ops_per_sec, read.latency_p50_ms, read.latency_p99_ms, read.wall_secs
    );

    // Keep store alive until footprint / reads done.
    drop(store);

    LayoutReport {
        layout: kind.as_str().into(),
        write,
        read,
        footprint,
    }
}

fn print_table(cfg: &Config, layouts: &[LayoutReport]) {
    println!();
    println!(
        "haystack_small  COUNT={}  SIZE={}  WARMUP={}  DROP_CACHES={}",
        cfg.count,
        human_bytes(cfg.size),
        cfg.warmup,
        cfg.drop_caches as u8
    );
    println!(
        "{:<10} {:>10} {:>8} {:>8} {:>10} {:>8} {:>8} {:>8} {:>8}",
        "layout", "w_ops/s", "w_p50", "w_p99", "r_ops/s", "r_p50", "r_p99", "obj_n", "vol_n"
    );
    println!("{}", "-".repeat(92));
    for l in layouts {
        println!(
            "{:<10} {:>10.1} {:>7.3}ms {:>7.3}ms {:>10.1} {:>7.3}ms {:>7.3}ms {:>8} {:>8}",
            l.layout,
            l.write.ops_per_sec,
            l.write.latency_p50_ms,
            l.write.latency_p99_ms,
            l.read.ops_per_sec,
            l.read.latency_p50_ms,
            l.read.latency_p99_ms,
            l.footprint.objects_files,
            l.footprint.volume_files,
        );
    }
    println!();
}

fn write_report(report: &Report) -> PathBuf {
    let out_dir = Path::new(env!("CARGO_MANIFEST_DIR")).join("bench/results");
    std::fs::create_dir_all(&out_dir).expect("mkdir results");
    let stamp = Utc::now().format("%Y%m%d-%H%M%S");
    let path = out_dir.join(format!("haystack_small-{stamp}.json"));
    let json = serde_json::to_vec_pretty(report).expect("serialize");
    std::fs::write(&path, json).expect("write report");
    path
}

#[tokio::main]
async fn main() {
    let cfg = Config::from_env();
    // Same rule as Haystack::open — banner only; open() is the source of truth.
    let haystack_max = std::env::var("HAYSTACK_MAX_VOLUME_SIZE")
        .ok()
        .and_then(|raw| raw.trim().parse().ok())
        .filter(|&n| n > 0)
        .unwrap_or(object_store::store::haystack::DEFAULT_MAX_VOLUME_SIZE);
    eprintln!(
        "haystack_small: count={} size={} warmup={} drop_caches={} haystack_max_volume={}",
        cfg.count,
        human_bytes(cfg.size),
        cfg.warmup,
        cfg.drop_caches,
        human_bytes(haystack_max),
    );
    eprintln!("(always use --release; see bench/haystack_small/README.md)\n");

    let mut layouts = Vec::new();
    for kind in [BlobLayoutKind::FileCas, BlobLayoutKind::Haystack] {
        layouts.push(run_layout(kind, &cfg).await);
        eprintln!();
    }

    print_table(&cfg, &layouts);
    let report = Report {
        created_at: Utc::now().to_rfc3339(),
        count: cfg.count,
        size_bytes: cfg.size,
        warmup: cfg.warmup,
        drop_caches: cfg.drop_caches,
        layouts,
    };
    let path = write_report(&report);
    eprintln!("wrote {}", path.display());
    eprintln!("curate the table into docs/06-benchmarks.md (Haystack vs FileCas).");
}

File: projects/06-object-store/bench/haystack_small/main.py
#!/usr/bin/env python3
"""FileCas vs Haystack for tiny objects — the small-object packing tradeoff.

Drives `Store.commit_bytes` / `Store.open_blob` directly, with no HTTP, so the
delta measured is physical layout — inodes, appends, ranged reads — and not the
router. See this directory's README for the honest-measurement checklist.

    make bench-haystack
    COUNT=20000 SIZE=4K uv run python bench/haystack_small/main.py
"""

from __future__ import annotations

import asyncio
import json
import os
import random
import shutil
import sys
import tempfile
import time
from dataclasses import dataclass, field
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2] / "src"))

from object_store.config import BlobLayoutKind  # noqa: E402
from object_store.objects import Digest  # noqa: E402
from object_store.store import Store  # noqa: E402

RESULTS = Path(__file__).resolve().parents[1] / "results"


def parse_size(raw: str) -> int:
    """`4K`, `1M`, `512` → bytes."""
    text = raw.strip().upper()
    for suffix, scale in (("K", 1024), ("M", 1024**2), ("G", 1024**3)):
        if text.endswith(suffix):
            return int(float(text[:-1]) * scale)
    return int(text)


@dataclass(slots=True)
class Phase:
    """Latency samples for one timed phase."""

    name: str
    samples: list[float] = field(default_factory=list)

    def record(self, seconds: float) -> None:
        self.samples.append(seconds)

    @property
    def ops_per_second(self) -> float:
        total = sum(self.samples)
        return len(self.samples) / total if total else 0.0

    def percentile(self, fraction: float) -> float:
        """Latency in milliseconds at `fraction`.

        Nearest-rank rather than an interpolating quantile: for a p99 over a few
                thousand samples an interpolated value can sit between two real
                measurements, which is fine for describing a distribution and misleading
                for a tail you are trying to explain.
        """
        if not self.samples:
            return 0.0
        ordered = sorted(self.samples)
        index = min(int(fraction * len(ordered)), len(ordered) - 1)
        return ordered[index] * 1000


def payload(index: int, size: int) -> bytes:
    """A unique payload per index.

    Stamping each one matters: identical bytes dedup, and a benchmark whose
    second half is all dedup hits measures the short-circuit, not the layout.
    """
    stamp = f"needle-{index:09d}-".encode()
    return stamp + bytes((index + position) % 256 for position in range(size - len(stamp)))


async def run_layout(
    layout: BlobLayoutKind, count: int, size: int, warmup: int, volume_cap: int
) -> dict[str, object]:
    root = Path(tempfile.mkdtemp(prefix=f"haystack-{layout.value}-"))
    try:
        store = Store(root, layout=layout, max_volume_size=volume_cap)

        for index in range(warmup):
            await store.commit_bytes(payload(-index - 1, size))

        writes = Phase("write")
        digests: list[Digest] = []
        for index in range(count):
            data = payload(index, size)
            started = time.perf_counter()
            digests.append(await store.commit_bytes(data))
            writes.record(time.perf_counter() - started)

        object_files = sum(1 for p in (root / "objects").rglob("*") if p.is_file())
        volume_files = sum(1 for p in (root / "volumes").glob("*.dat"))
        on_disk = sum(
            p.stat().st_size
            for name in ("objects", "volumes")
            for p in (root / name).rglob("*")
            if p.is_file()
        )

        # Shuffled so the read phase cannot ride sequential locality that a real
        # workload would not have.
        reads = Phase("read")
        random.Random(7).shuffle(digests)
        for digest in digests:
            started = time.perf_counter()
            reader = await store.open_blob(digest)
            try:
                while reader.read(64 * 1024):
                    pass
            finally:
                reader.close()
            reads.record(time.perf_counter() - started)

        store.close()
        return {
            "layout": layout.value,
            "write_ops_per_sec": round(writes.ops_per_second, 1),
            "write_p50_ms": round(writes.percentile(0.50), 3),
            "write_p99_ms": round(writes.percentile(0.99), 3),
            "read_ops_per_sec": round(reads.ops_per_second, 1),
            "read_p50_ms": round(reads.percentile(0.50), 3),
            "read_p99_ms": round(reads.percentile(0.99), 3),
            "object_files": object_files,
            "volume_files": volume_files,
            "bytes_on_disk": on_disk,
        }
    finally:
        shutil.rmtree(root, ignore_errors=True)


async def main() -> None:
    count = int(os.environ.get("COUNT", "5000"))
    size = parse_size(os.environ.get("SIZE", "4K"))
    warmup = int(os.environ.get("WARMUP", "100"))
    # 1 GiB by default here, unlike the store's 1 MiB: thousands of 4 KiB
    # needles must land in *one* volume or the comparison measures volume
    # rollover instead of packing.
    volume_cap = parse_size(os.environ.get("HAYSTACK_MAX_VOLUME_SIZE", "1G"))

    print(f"count={count} size={size}B warmup={warmup} volume_cap={volume_cap}B\n")

    rows = [
        await run_layout(layout, count, size, warmup, volume_cap)
        for layout in (BlobLayoutKind.FILE_CAS, BlobLayoutKind.HAYSTACK)
    ]

    header = (
        f"{'layout':<10} {'w_ops/s':>10} {'w_p50':>8} {'w_p99':>8} "
        f"{'r_ops/s':>10} {'r_p50':>8} {'r_p99':>8} {'obj_n':>8} {'vol_n':>7}"
    )
    print(header)
    print("-" * len(header))
    for row in rows:
        print(
            f"{row['layout']:<10} {row['write_ops_per_sec']:>10} "
            f"{row['write_p50_ms']:>8} {row['write_p99_ms']:>8} "
            f"{row['read_ops_per_sec']:>10} {row['read_p50_ms']:>8} "
            f"{row['read_p99_ms']:>8} {row['object_files']:>8} {row['volume_files']:>7}"
        )

    RESULTS.mkdir(parents=True, exist_ok=True)
    out = RESULTS / f"haystack_small-{int(time.time())}.json"
    out.write_text(
        json.dumps(
            {"count": count, "size": size, "volume_cap": volume_cap, "rows": rows},
            indent=2,
        )
    )
    print(f"\n→ {out}")
    print("Curate the table you care about into docs/06-benchmarks.md.")


if __name__ == "__main__":
    asyncio.run(main())
